build/common.sh

```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
-      fi
+  local container_ip
+  container_ip=$("${DOCKER[@]}" inspect --format '{{ .NetworkSettings.IPAddress }}' "${KUBE_RSYNC_CONTAINER_NAME}")
+
+  if kube::build::probe_address 127.0.0.1 ${mapped_port}; then
+    KUBE_RSYNC_ADDR="127.0.0.1:${mapped_port}"
+    sleep 0.5
+    return 0
+  elif kube::build::probe_address "${container_ip}" 8730; then
+    KUBE_RSYNC_ADDR="${container_ip}:8730"
+    sleep 0.5
+    return 0
+  fi
 
-      let count=count+1
-      if [[ $count -eq 60 ]]; then
-        # break after 5m
-        kube::log::error "Timed out waiting for binaries..."
-        break
-      fi
-      sleep 5
-    done
+  kube::log::error "Could not connect to rsync container. See build/README.md for setting up remote Docker engine."
+  return 1
+}
 
-    "${DOCKER[@]}" rm -f -v "${KUBE_BUILD_CONTAINER_NAME}" >/dev/null 2>&1 || true
-  else
-    kube::log::status "Output directory is local.  No need to copy results out."
+function kube::build::stop_rsyncd_container() {
+  V=6 kube::log::status "Stopping any currently running rsyncd container"
+  kube::build::destroy_container "${KUBE_RSYNC_CONTAINER_NAME}"
+}
+
+# This will launch rsyncd in a container and then sync the source tree to the
+# container over the local network.
+function kube::build::sync_to_container() {
+  kube::log::status "Syncing sources to container"
+
+  kube::build::start_rsyncd_container
+
+  local rsync_extra=""
+  if (( $KUBE_VERBOSE >= 6 )); then
+    rsync_extra="-iv"
+  fi
+
+  V=6 kube::log::status "Running rsync"
+  rsync $rsync_extra \
+    --archive \
+    --prune-empty-dirs \
+    --password-file="${LOCAL_OUTPUT_BUILD_CONTEXT}/rsyncd.password" \
+    --filter='- /.git/' \
+    --filter='- /.make/' \
+    --filter='- /_tmp/' \
+    --filter='- /_output/' \
+    --filter='- /' \
+    "${KUBE_ROOT}/" "rsync://k8s@${KUBE_RSYNC_ADDR}/k8s/"
+
+  kube::build::stop_rsyncd_container
+}
+
+# If the Docker server is remote, copy the results back out.
+function kube::build::copy_output() {
+  kube::log::status "Syncing out of container"
+
+  kube::build::start_rsyncd_container
+
+  local rsync_extra=""
+  if (( $KUBE_VERBOSE >= 6 )); then
+    rsync_extra="-iv"
   fi
+
+  # The filter syntax for rsync is a little obscure. It filters on files and
+  # directories.  If you don't go in to a directory you won't find any files
+  # there.  Rules are evaluated in order.  The last two rules are a little
+  # magic. '+ */' says to go in to every directory and '- /**' says to ignore
+  # any file or directory that isn't already specifically allowed.
+  #
+  # We are looking to copy out all of the built binaries along with various
+  # generated files.
+  V=6 kube::log::status "Running rsync"
+  rsync $rsync_extra \
+    --archive \
+    --prune-empty-dirs \
+    --password-file="${LOCAL_OUTPUT_BUILD_CONTEXT}/rsyncd.password" \
+    --filter='- /vendor/' \
+    --filter='- /_temp/' \
+    --filter='+ /_output/dockerized/bin/**' \
+    --filter='+ zz_generated.*' \
```

Comment:
I don't really have a better idea (besides possibly having some process which notices changes to the volume?), but we might want to open an issue to keep track of this. having to keep a list of everything we might ever generate in random places is pretty bad.